lib/MySQL_Monitor.cpp

```diff
@@ -5816,5 +5856,1433 @@ bool DNS_Cache::empty() const {
+				}
+			}
+
+			for (unsigned int i = 0; i < len_;) {
+
+				if (fds_[i].revents != 0) {
+
+					if (mmsds_[i]->task_handler(fds_[i].revents, fds_[i].events) == MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_PENDING) {
+						assert(fds_[i].events != 0);
+					}
+					else {
+						remove_index_fast(i);
+						continue;
+					}
+				}
+				fds_[i].revents = 0;
+				i++;
+			}
+
+			t1 = t2;
+			t2 = monotonic_time()/1000;
+			poll_timeout_ms -= static_cast<int>(t2 - t1);
+
+			if (poll_timeout_ms <= 0) {
+				break;
+			}
+		}
+
+		for (unsigned int i = 0; i < len_; i++) {
+			mmsds_[i]->mark_task_as_timeout();
+		}
+
+		return true;
+	}
+
+	inline 
+	unsigned int count() const {
+		return len_;
+	}
+
+private:
+	static unsigned int near_pow_2(unsigned int n) {
+		unsigned int i = 1;
+		while (i < n) i <<= 1;
+		return i ? i : n;
+	}
+
+	unsigned int len_;
+	unsigned int size_;
+	struct pollfd* fds_;
+	MySQL_Monitor_State_Data** mmsds_;
+};
+
+MySQL_Monitor_State_Data_Task_Result MySQL_Monitor_State_Data::ping_handler(short event_, short& wait_event) {
+	MySQL_Monitor_State_Data_Task_Result result = MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_PENDING;
+	int status = 0;
+
+__again:
+	proxy_debug(PROXY_DEBUG_MYSQL_PROTOCOL, 6, "async_state_machine=%d\n", async_state_machine_);
+	switch (async_state_machine_) {
+	case ASYNC_PING_START:
+		t1 = monotonic_time();
+		interr = 0;
+		if (mysql_error_msg) {
+			free(mysql_error_msg);
+			mysql_error_msg = NULL;
+		}
+		status = mysql_ping_start(&interr, mysql);
+		if (status) {
+			wait_event = next_event(ASYNC_PING_CONT, status);
+		}
+		else {
+			NEXT_IMMEDIATE(ASYNC_PING_END);
+		}
+		break;
+	case ASYNC_PING_CONT:
+		status = mysql_ping_cont(&interr, mysql, mysql_status(event_));
+	
+		if (status) {
+			unsigned long long now = monotonic_time();
+			if (now > t1 + timeout_ * 1000) {
+				t2 = now;
+				mysql_error_msg = strdup("timeout during ping");
+				NEXT_IMMEDIATE(ASYNC_PING_TIMEOUT);
+			}
+
+			wait_event = next_event(ASYNC_PING_CONT, status);
+		}
+		else {
+			NEXT_IMMEDIATE(ASYNC_PING_END);
+		}
+		break;
+	case ASYNC_PING_END:
+		t2 = monotonic_time();
+		if (interr) {
+			mysql_error_msg = strdup(mysql_error(mysql));
+			NEXT_IMMEDIATE(ASYNC_PING_FAILED);
+		}
+		else {
+			NEXT_IMMEDIATE(ASYNC_PING_SUCCESSFUL);
+		}
+		break;
+	case ASYNC_PING_SUCCESSFUL:
+		result = MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_SUCCESS;
+		break;
+	case ASYNC_PING_FAILED:
+		result = MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_FAILED;
+		break;
+	case ASYNC_PING_TIMEOUT:
+		//t2 = monotonic_time();
+		result = MySQL_Monitor_State_Data_Task_Result::TASK_RESULT_TIMEOUT;
+		break;
+	default:
+		assert(0);
+		break;
+	}
+
+	return result;
+}
+
+void MySQL_Monitor::monitor_ping_async(SQLite3DB* monitordb, SQLite3_result* resultset) {
+
+	Monitor_Poll monitor_poll;
+	std::vector<std::unique_ptr<MySQL_Monitor_State_Data>> mmsds;
+
+	mmsds.reserve(resultset->rows_count);
```

Comment:
Some tasks will be delegated to consumer thread some will be resolved asynchronously. We only need to track async tasks so that's why reserving max memory here.